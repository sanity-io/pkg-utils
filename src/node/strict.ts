import {z} from 'zod'
import {errorMap} from 'zod-validation-error'

/**
 * @public
 */
export const toggle = z.union([z.literal('error'), z.literal('warn'), z.literal('off')])

/**
 * @public
 */
export type ToggleType = z.infer<typeof toggle>

/**
 * @public
 */
export const strictOptions = z
  .object({
    noPackageJsonTypings: toggle.default('error'),
    noImplicitSideEffects: toggle.default('warn'),
    noImplicitBrowsersList: toggle.default('warn'),
    alwaysPackageJsonTypes: toggle.default('error'),
    alwaysPackageJsonMain: toggle.default('error'),
    alwaysPackageJsonFiles: toggle.default('error'),
  })
  .strict()

/**
 * To make error message paths line up with the paths in package.config.ts the schema is hoisted into a root schema
 * This way errors will say `Expected boolean, received string at "strict.noPackageJsonTypings"` instead of `Expected boolean, received string at "noPackageJsonTypings"`.
 */
const validationSchema = z.object({
  strictOptions: strictOptions.default({}),
})

/**
 * @public
 */
export type InferredStrictOptions = z.infer<typeof strictOptions>

/**
 * @public
 */
export interface StrictOptions {
  /**
   * Disallows a top level `typings` field in `package.json` if it is equal to `exports['.'].source`.
   * @defaultValue 'error'
   */
  noPackageJsonTypings?: ToggleType
  /**
   * Requires specifying `sideEffects` in `package.json`.
   * @defaultValue 'warn'
   */
  noImplicitSideEffects?: ToggleType
  /**
   * Requires specifying `browserslist` in `package.json`, instead of relying on it implicitly being:
   * @example
   * ```
   * "browserslist": "extends @sanity/browserslist-config"
   * ```
   * @defaultValue 'warn'
   */
  noImplicitBrowsersList?: ToggleType
  /**
   * If typescript is used then `types` in `package.json` should be specified for npm listings to show the TS icon.
   * @defaultValue 'error'
   */
  alwaysPackageJsonTypes?: ToggleType
  /**
   * A lot of analysis tooling requiers the `main` field to work (like bundlephobia) and so it's best practice to always include it
   * @defaultValue 'error'
   */
  alwaysPackageJsonMain?: ToggleType
  /**
   * Using `.npmignore` is error prone, it's best practice to always declare `files` instead
   * @defaultValue 'error'
   */
  alwaysPackageJsonFiles?: ToggleType
}

/** @internal */
export function parseStrictOptions(input: unknown): InferredStrictOptions {
  return validationSchema.parse({strictOptions: input}, {errorMap}).strictOptions
}
