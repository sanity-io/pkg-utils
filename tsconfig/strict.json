{
  "$schema": "https://json.schemastore.org/tsconfig",
  "description": "Extends the recommended preset with stricter settings.",
  "extends": "./recommended.json",
  "compilerOptions": {
    "checkJs": true,
    "isolatedModules": true,
    "verbatimModuleSyntax": true,
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true,
    "noPropertyAccessFromIndexSignature": true
  }
}
