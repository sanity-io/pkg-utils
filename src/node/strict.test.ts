import {expect, test} from 'vitest'
import {parseStrictOptions} from './strict'

test.each([
  {key: 'noPackageJsonTypings', value: 'error', fails: false},
  {key: 'noPackageJsonTypings', value: 'warn', fails: false},
  {key: 'noPackageJsonTypings', value: 'off', fails: false},
  {key: 'noPackageJsonTypings', value: true, fails: true},
  {key: 'noPackageJsonTypings', value: false, fails: true},
])('{$key: $value}}', ({key, value, fails}) => {
  if (fails) {
    expect(() => parseStrictOptions({[key]: value})).toThrowErrorMatchingSnapshot()
  } else {
    expect(parseStrictOptions({[key]: value})).toMatchSnapshot()
  }
})
