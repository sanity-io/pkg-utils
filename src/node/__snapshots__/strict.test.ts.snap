// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`{'noPackageJsonTypings': 'error'}} 1`] = `
{
  "alwaysPackageJsonFiles": "error",
  "alwaysPackageJsonMain": "error",
  "alwaysPackageJsonTypes": "error",
  "noImplicitBrowsersList": "warn",
  "noImplicitSideEffects": "warn",
  "noPackageJsonTypings": "error",
}
`;

exports[`{'noPackageJsonTypings': 'off'}} 1`] = `
{
  "alwaysPackageJsonFiles": "error",
  "alwaysPackageJsonMain": "error",
  "alwaysPackageJsonTypes": "error",
  "noImplicitBrowsersList": "warn",
  "noImplicitSideEffects": "warn",
  "noPackageJsonTypings": "off",
}
`;

exports[`{'noPackageJsonTypings': 'warn'}} 1`] = `
{
  "alwaysPackageJsonFiles": "error",
  "alwaysPackageJsonMain": "error",
  "alwaysPackageJsonTypes": "error",
  "noImplicitBrowsersList": "warn",
  "noImplicitSideEffects": "warn",
  "noPackageJsonTypings": "warn",
}
`;

exports[`{'noPackageJsonTypings': false}} 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": false,
            "code": "invalid_literal",
            "expected": "error",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"error\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": false,
            "code": "invalid_literal",
            "expected": "warn",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"warn\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": false,
            "code": "invalid_literal",
            "expected": "off",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"off\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "strictOptions",
      "noPackageJsonTypings"
    ],
    "message": "Validation error: Validation error: Invalid literal value, expected \\"error\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\", or Validation error: Invalid literal value, expected \\"warn\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\", or Validation error: Invalid literal value, expected \\"off\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\""
  }
]]
`;

exports[`{'noPackageJsonTypings': true}} 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": true,
            "code": "invalid_literal",
            "expected": "error",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"error\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": true,
            "code": "invalid_literal",
            "expected": "warn",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"warn\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": true,
            "code": "invalid_literal",
            "expected": "off",
            "path": [
              "strictOptions",
              "noPackageJsonTypings"
            ],
            "message": "Validation error: Invalid literal value, expected \\"off\\" at \\"strictOptions.noPackageJsonTypings\\""
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "strictOptions",
      "noPackageJsonTypings"
    ],
    "message": "Validation error: Validation error: Invalid literal value, expected \\"error\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\", or Validation error: Invalid literal value, expected \\"warn\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\", or Validation error: Invalid literal value, expected \\"off\\" at \\"strictOptions.noPackageJsonTypings\\" at \\"strictOptions.noPackageJsonTypings\\""
  }
]]
`;
